spec/unit/git_resolver_spec.cr

```diff
@@ -28,6 +28,26 @@ module Shards
       create_git_tag "library", "99.9.9"
     end
 
+    it "normalizes github bitbucket gitlab sources" do
+      # deal with case insensitive paths
+      paths = {"repo/path", "rEpo/pAth", "REPO/PATH"}
+      keys = {"github", "bitbucket", "gitlab"}
+      results = {"https://github.com/repo/path.git", "https://bitbucket.com/repo/path.git", "https://gitlab.com/repo/path.git"}
+      keys.each_with_index do |key, index|
+        paths.each { |path| GitResolver.normalize_key_source(key, path).should eq({"git", results[index]}) }
+      end
+
+      # also normalise full git paths
+      paths = {"HTTPS://User:Pass@Github.com/Repo/Path.git?Shallow=true", "HTTPS://User:Pass@Bitbucket.com/Repo/Path.Git?Shallow=true", "HTTPS://User:Pass@Gitlab.com/Repo/Path?Shallow=true"}
+      results = {"https://User:Pass@github.com/repo/path.git?Shallow=true", "https://User:Pass@bitbucket.com/repo/path.git?Shallow=true", "https://User:Pass@gitlab.com/repo/path.git?Shallow=true"}
+      paths.each_with_index do |path, index|
+        GitResolver.normalize_key_source("git", path).should eq({"git", results[index]})
+      end
```

Comment:
Such programmatical expanded code makes the examples unnecessarily hard to understand. This is overengineering. If you just use plain expecations, it much more clearly expresses what's going on.
```suggestion
      GitResolver.normalize_key_source("github", "repo/path").should eq({"git", "https://github.com/repo/path"})
      GitResolver.normalize_key_source("github", "rEpo/pAth").should eq({"git", "https://github.com/repo/path"})
      GitResolver.normalize_key_source("github", "REPO/PATH").should eq({"git", "https://github.com/repo/path"})
      GitResolver.normalize_key_source("bitbucket", "repo/path").should eq({"git", "https://bitbucket.com/repo/path"})
      # GitResolver.normalize_key_source("bitbucket", "rEpo/pAth").should eq({"git", "https://bitbucket.com/repo/path"})
      # GitResolver.normalize_key_source("bitbucket", "REPO/PATH").should eq({"git", "https://bitbucket.com/repo/path"})
      GitResolver.normalize_key_source("gitlab", "repo/path").should eq({"git", "https://gitlab.com/repo/path"})
      # GitResolver.normalize_key_source("gitlab", "rEpo/pAth").should eq({"git", "https://gitlab.com/repo/path"})
      # GitResolver.normalize_key_source("gitlab", "REPO/PATH").should eq({"git", "https://gitlab.com/repo/path"})

      # also normalise full git paths
      GitResolver.normalize_key_source("git", "HTTPS://User:Pass@Github.com/Repo/Path.git?Shallow=true").should eq "https://User:Pass@github.com/repo/path.git?Shallow=true"
      GitResolver.normalize_key_source("git", "HTTPS://User:Pass@Bitbucket.com/Repo/Path.Git?Shallow=true"). should eq "https://User:Pass@bitbucket.com/repo/path.git?Shallow=true"
      GitResolver.normalize_key_source("git", "HTTPS://User:Pass@Gitlab.com/Repo/Path?Shallow=true").should eq "https://User:Pass@gitlab.com/repo/path.git?Shallow=true"
```

We probably don't need full validation for every combination between path style and host. As long as the behaviour is identical beween hosts, skipping a couple of the duplicated tests is fine (I've included them in the diff but commented out; they can be removed).

This is not a requirement, but I'd really prefer this.